src/project.rs

```diff
@@ -0,0 +1,191 @@
+use crate::docker::{ImageArchUri, DEFAULT_REGISTRY, DEFAULT_SDK_NAME, DEFAULT_SDK_VERSION};
+use anyhow::{ensure, Context, Result};
+use async_recursion::async_recursion;
+use log::trace;
+use serde::de::Error;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::fmt;
+use std::path::{Path, PathBuf};
+use tokio::fs;
+
+/// Represents the structure of a `Twoliter.toml` project file.
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+pub(crate) struct Project {
+    pub(crate) schema_version: SchemaVersion<1>,
+    pub(crate) project_name: String,
+    pub(crate) project_version: String,
+    pub(crate) sdk: Option<Sdk>,
+}
+
+impl Project {
+    /// Load a `Twoliter.toml` file from the given file path (it can have any filename).
+    pub(crate) async fn load<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let path = path.as_ref();
+        let data = fs::read_to_string(&path)
+            .await
+            .context(format!("Unable to read project file '{}'", path.display()))?;
+        toml::from_str(&data).context(format!(
+            "Unable to deserialize project file '{}'",
+            path.display()
+        ))
+    }
+
+    /// Recursively search for a file named `Twoliter.toml` starting in `dir`. If it is not found,
+    /// move up (i.e. `cd ..`) until it is found. Return an error if there is no parent directory.
+    #[async_recursion]
+    pub(crate) async fn find_and_load<P: AsRef<Path> + Send>(dir: P) -> Result<(Self, PathBuf)> {
+        let dir = dir.as_ref();
+        trace!("Looking for Twoliter.toml in '{}'", dir.display());
+        ensure!(
+            dir.is_dir(),
+            "Unable to locate Twoliter.toml in '{}': not a directory",
+            dir.display()
+        );
+        let dir = dir
+            .canonicalize()
+            .context(format!("Unable to canonicalize '{}'", dir.display()))?;
+        let filepath = dir.join("Twoliter.toml");
+        if filepath.is_file() {
+            return Ok((Self::load(&filepath).await?, filepath));
+        }
+        // Move up a level and recurse.
+        let parent = dir
+            .parent()
+            .context("Unable to find Twoliter.toml file")?
+            .to_owned();
```

Comment:
Potentially we could stop as soon as we reach a directory that has a `.git` subdirectory, since beyond that point it's unlikely we're going to find the `Twoliter.toml` that the user expects.